pkg/gardenlet/controller/federatedseed/shootstate/common_sync.go

```diff
@@ -0,0 +1,156 @@
+			return
+		}
+		enqueue(queue, newObj)
+	}
+}
+
+func enqueue(queue workqueue.RateLimitingInterface, obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+	queue.Add(key)
+}
+
+func (s *SyncController) createShootStateSyncReconcileFunc(ctx context.Context, kind string, obj runtime.Object) reconcile.Func {
+	return func(req reconcile.Request) (reconcile.Result, error) {
+		obj = obj.DeepCopyObject()
+		err := s.seedClient.Client().Get(ctx, req.NamespacedName, obj)
+		if apierrors.IsNotFound(err) {
+			s.log.Debugf("Skipping ShootState sync because resoruce with kind %s is missing", kind)
+			return reconcile.Result{}, nil
+		}
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		extensionObject, err := apiextensions.Accessor(obj)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		state := extensionObject.GetExtensionStatus().GetState()
+		if state == nil {
+			return reconcile.Result{}, fmt.Errorf("invalid state. Extension state for resource with kind %s in namespace %s was changed to nil before it could be synced to ShootState", kind, req.Namespace)
+		}
+
+		name := extensionObject.GetName()
+		purpose := getPurpose(obj)
+		projectNamespace, shootName := fromNamespaceName(req.Namespace)
+		shootState := &gardencorev1alpha1.ShootState{ObjectMeta: metav1.ObjectMeta{Name: shootName, Namespace: projectNamespace}}
+		if _, err := controllerutil.CreateOrUpdate(ctx, s.k8sGardenClient.Client(), shootState, func() error {
+			return updateShootStateExtensionState(state, shootState, kind, name, purpose)
+		}); err != nil {
+			message := fmt.Sprintf("Shoot %s extension state was NOT succcessfuly synced", kind)
+			s.log.Error(message)
+			s.recorder.Event(shootState, corev1.EventTypeNormal, "ScheduledNextSync", message)
+			return reconcile.Result{}, err
+		}
+
+		message := fmt.Sprintf("Shoot %s extension state was succcessfuly synced", kind)
+		s.log.Info(message)
+		s.recorder.Event(shootState, corev1.EventTypeNormal, "ScheduledNextSync", message)
+		return reconcile.Result{Requeue: false}, nil
+	}
+}
+
+func updateShootStateExtensionState(extensionState *runtime.RawExtension, shootState *gardencorev1alpha1.ShootState, kind string, name string, purpose *string) error {
+	currentExtensionState := gardencorev1alpha1helper.GetExtensionResourceState(shootState.Spec.Extensions, kind, &name, purpose)
+	if currentExtensionState == nil {
+		shootState.Spec.Extensions = append(shootState.Spec.Extensions, gardencorev1alpha1.ExtensionResourceState{
+			Kind:    kind,
+			Name:    &name,
+			Purpose: purpose,
+			State:   gardencorev1alpha1.ProviderConfig{RawExtension: *extensionState},
+		})
+	} else {
+		currentExtensionState.State = gardencorev1alpha1.ProviderConfig{RawExtension: *extensionState}
+	}
+	return nil
+}
+
+func getPurpose(obj runtime.Object) *string {
```

Comment:
Does it make sense to move this to https://github.com/gardener/gardener/blob/master/pkg/api/extensions/accessor.go#L144? It would return the purpose for `ControlPlane`s and `OperatingSystemConfig`s, and `""` for all other types (because the `purpose` field does not exist there).

Reply:
We could add such a function, however the Purpose field for the OperatingSystemConfig and the ControlPlane are different types `string` and `*string` so i guess it would be better to let the GetPurpose function also return *string.